motd/src/reimpl.rs

```diff
@@ -0,0 +1,151 @@
+                if let Some(motd_dir_arg) = arg.strip_prefix("motd_dir=") {
+                    motd_d = motd_dir_arg;
+                }
+            } else if arg != "noupdate" {
+                warn!("unknown motd arg '{}', ignoring", arg);
+            }
+        }
+
+        let mut msg = String::from("");
+        self.slurp_motd(motd, &mut msg)?;
+        self.slurp_motd_d(motd_d, &mut msg)?;
+
+        Ok(msg)
+    }
+
+    /// Go through the motd list and slurp the first file that exists.
+    fn slurp_motd(&self, motd: &str, into: &mut String) -> Result<(), Error> {
+        for path in motd.split(':') {
+            if !fs::metadata(path).is_ok() {
+                // motd file not present, just skip it
+                continue;
+            }
+            let contents =
+                fs::read_to_string(path).map_err(|e| merr!("reading motd file: {:?}", e))?;
+            into.push_str(contents.as_str());
+            break;
+        }
+
+        Ok(())
+    }
+
+    // Go through the motd_d list and slurp all the files, applying
+    // name based overrides as described in `man pam_motd`.
+    fn slurp_motd_d(&self, motd_d: &str, into: &mut String) -> Result<(), Error> {
+        let mut direntries = vec![];
+        let mut files = vec![];
+        for dir in motd_d.split(':') {
+            let entries = match fs::read_dir(dir) {
+                Ok(e) => e,
+                Err(e) => {
+                    // just warn about it since the next one might be a real
+                    // directory.
+                    warn!("reading motd_d dir: {:?}", e);
+                    continue;
+                }
+            };
+
+            let mut entry_list = vec![];
+            for entry in entries {
+                let entry = entry.map_err(|e| merr!("resolving dir entry: {:?}", e))?;
+                if entry.path().is_symlink() || entry.path().is_file() {
+                    files.push(
+                        entry
+                            .path()
+                            .file_name()
+                            .and_then(|s| s.to_str())
+                            .map(|s| String::from(s))
+                            .ok_or(merr!("could not get basename"))?,
+                    );
+                }
+                entry_list.push(entry);
+            }
+
+            direntries.push(entry_list);
+        }
+
+        files.sort();
+
+        for (i, file) in files.iter().enumerate() {
+            if i > 0 && files[i - 1] == file.as_str() {
+                // skip dups
+                continue;
+            }
+
+            for entries in direntries.iter() {
+                for entry in entries {
+                    // skip non-matching files
+                    if !entry
+                        .path()
+                        .file_name()
+                        .and_then(|s| s.to_str())
+                        .map(|s| s == file)
+                        .unwrap_or(false)
```

Comment:
is_some(), or better drop the ! and use is_none()

Reply:
Well I need more info than just if the Option is occupied or not. I want to know if the filename actually matches for which I need some sort of unwrap. This is saying "if the files match, where a missing path counts as not a match" and then I negate the whole thing. I think that's clearer than inverting the == and doing .unwrap_or(true).